LiteCore/Support/Logging.cc

```diff
@@ -55,36 +55,131 @@ namespace litecore {
     static LogDomain::Callback_t sCallback = LogDomain::defaultCallback;
     static bool sCallbackPreformatted = false;
     LogLevel LogDomain::sFileMinLevel = LogLevel::None;
-    static ofstream *sFileOut = nullptr;
-    static LogEncoder* sLogEncoder = nullptr;
+    static ofstream* sFileOut[5] = {}; // File per log level
+    static LogEncoder* sLogEncoder[5] = {};
+    static string sLogDirectory;
+    static int sMaxCount = 0;       // For rotation
+    static int64_t sMaxSize = 1024; // For rotation
     static mutex sLogMutex;
-    static const int kNumLogFiles = 10;
 
+    static const char* const kLevelNames[] = {"debug", "verbose", "info",
+                "warning", "error", nullptr};
+    static const char *kLevels[] = {"***", "", "", "WARNING", "ERROR"};
+
+    static string createLogPath(LogLevel level)
+    {
+        time_t result = time(nullptr);
+
+        stringstream ss;
+        ss << sLogDirectory << FilePath::kSeparator << "cbl_" << kLevelNames[(int)level] << "_" << result;
+        return ss.str();
+    }
+
+    static void setupFileOut()
+    {
+        for (int i = 0; kLevelNames[i]; i++) {
+            auto path = createLogPath((LogLevel)i);
+            sFileOut[i] = new ofstream(path, ofstream::out|ofstream::trunc|ofstream::binary);
+        }
+    }
+
+    static void setupEncoders()
+    {
+        for(int i = 0; i < 5; i++) {
+            sLogEncoder[i]  = new LogEncoder(*sFileOut[i], (LogLevel)i);
+        }
+    }
+
+    static void teardownEncoders()
+    {
+        for(auto& encoder : sLogEncoder) {
+            if(encoder) encoder->flush();
+            delete encoder;
+            encoder = nullptr;
+        }
+    }
+
+    static void teardownFileOut()
+    {
+        for (auto& fout : sFileOut) {
+            if(fout) fout->flush();
+            delete fout;
+            fout = nullptr;
+        }
+    }
+
+    static bool needsTeardown(const LogFileOptions& options)
+    {
+        if(sLogEncoder[0] == nullptr && !options.isPlaintext()) {
+            return true;
+        }
+
+        if(sLogEncoder[0] != nullptr && options.isPlaintext()) {
+            return true;
+        }
+
+        if(sLogDirectory != options.path()) {
+            return true;
+        }
+
+        return false;
+    }
 
-    static void purgeOldLogs(const string& filePath)
+    static void purgeOldLogs(LogLevel level)
     {
-        FilePath filePathObj(filePath);
-        FilePath logDir = filePathObj.dir();
-        string ext = filePathObj.extension();
+        FilePath logDir(sLogDirectory);
         multimap<time_t, FilePath> logFiles;
+        const char* levelStr = kLevelNames[(int)level];
+
         logDir.forEachFile([&](const FilePath& f) {
-            if (f.extension() == ext) {
-                char magicBuffer[4] = {};
+            if (f.fileName().find(levelStr) != string::npos) {
```

Comment:
Why not just use a custom extension like `.cbllog`? Then you wouldn't have to be paranoid and check the magic number in the file.

Reply:
I would still be paranoid that someone for some reason had inserted a file with the same extension for the lolz.  It would then be deleted and it would be our fault.  Is that overboard?

Reply:
Yeah, that's overboard ;-)